Updated handling of arguments

dresdor opened the pull request:
Updated handling of arguments to avoid issues with spaces.

dresdor committed:
Updated handling of arguments

Updated handling of arguments to avoid issues with spaces.

smudge202 commented:
I hadn't seen this pattern before so I admit it was confusing at first. I do like it now I understand how it's doing the odd/even parsing.

Curious... what happens to the pattern if we were to add a flag? For example, if we added `-v|--verbose`, what would happen to this pattern?

dresdor commented:
if you were to add a flag that wasn't proceeded by the double dash, you'd need to update line 49 to do a .contains on "-" instead of "--".  To add any other flag, you'd need to add it to the dictionary in 32-38.  It may be better to define the dictionary as a class variable from the beginning, but I didn't check to see if the variables that went into it were used anywhere else.

If you are referring to flags that do not use any paired parameter with them, then there will definitely need rework.  Quick fix there is to call them out specifically in code (if there's just one) when parsing arguments and setting their dictionary value to true.  If there are a lot of boolean flags in your arguments, you should instead define a second dictionary for them since there will be separate behavior.

stajs commented:
@dresdor first of all, thanks for the submission. I have a couple of notes.

Spaces are stripped from arguments but that isn't always the best thing to do:

![image](https://user-images.githubusercontent.com/2253814/35141703-b8bf9ca8-fd60-11e7-8eb2-f44c24afaf97.png)

This is a well-known pathing problem with the usual solution of double-quoting being acceptable:

![image](https://user-images.githubusercontent.com/2253814/35141850-4bfecaa2-fd61-11e7-90fe-b60e438fbb58.png)

Unless you happen to have a trailing slash on your path:

![image](https://user-images.githubusercontent.com/2253814/35141866-60a36cb0-fd61-11e7-8a93-63f88661bc1f.png)

It's no worse than the previous unpredictable behaviour so I can live with this. I have a few other things that I'll add on the code review.

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -1,79 +1,91 @@
+ 
+ 			string lastKey = null;
+ 			// loop through all arguments, attempting to detect and fix bad values
+ 			for (var x = 0; x<args.Length; x++)
```
Nit: It's idiomatic to use ` i` for variable name.

dresdor replied:
Updated to be a foreach.

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -1,79 +1,91 @@
+ 				{
+ 					lastKey = args[x];
+ 				}
+				else if (args[x].Contains("--") || string.IsNullOrEmpty(lastKey))
```
Nit: this first check should be using `StartsWith()` to match the comment.

dresdor replied:
updated

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -1,79 +1,91 @@
+ 					case WorkingDirectoryArgName:
+ 						if (!Directory.Exists(argDictionary[key]))
+ 							throw new Exception("Working directory doesn't exist: " + argDictionary[key]);
+ 						WorkingDirectory = new DirectoryInfo(argDictionary[key]);
```
This needs to be wrapped in a null check and only update the `WorkingDirectory` property if the arg has been supplied, otherwise it should continue to use the current directory as default. If you supply any other argument but not the `--working-directory`,  `argDictionary["--working-directory"]` remains `null` failing the `Directory.Exists()`.

![image](https://user-images.githubusercontent.com/2253814/35142753-23d9eff4-fd64-11e7-8dcd-aa8053909797.png)

dresdor replied:
updated

stajs reviewed (changes requested):
The `WorkingDirectory` property should use the default value if not supplied.

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -1,79 +1,91 @@
+ 
+ 			WriteLine("SpecFlowPath: " + SpecFlowPath);
+ 			WriteLine("WorkingDirectory: " + WorkingDirectory.FullName);
+ 			WriteLine("TestFramework: " + TestFramework);
```
Note to self: add the `ToolsVersion` too.

dresdor committed:
updated from code review

updated from code review suggestions

dresdor commented:
Updated....it's supposed to pull the new code over automatically when I push to my fork but I don't see the update.  Not sure if it is delayed or I'm just missing it.

EDIT: Of course as I post this it appears.  Github is trolling me.

dresdor commented:
I was thinking about the spacing issue.  Should we implicitly add a space if two args are combined?  Wouldn't that be a valid assumption?  That should cover most cases.  However, it wouldn't cover if more than one space was needed in a path but that seems like an outlier.

Oddly enough the spacing issue was the whole reason I did this pull request and I didn't even fix it >_<   Our pre compiled command used html encoded quotes and didn't have an issue (but it took forever to figure that out, we kept getting the uneven arguments error).

dresdor committed:
added space to combined args

args tokenizes on spaces, so if two args are combined there should be a space between them.

stajs commented:
As I said, it's acceptable having to double-quote paths (or any arguments with spaces). And I had to laugh at the "pull request didn't even fix it"!

stajs commented on src/SpecFlow.NetCore/Args.cs:
```diff
@@ -21,59 +22,74 @@ public Args(string[] args)
-						if (!Directory.Exists(value))
-							throw new Exception("Working directory doesn't exist: " + value);
-						WorkingDirectory = new DirectoryInfo(value);
+						var path = string.IsNullOrEmpty(argDictionary[key]) ? Directory.GetCurrentDirectory(). : argDictionary[key];
```
The errant `.` here in `Directory.GetCurrentDirectory().` is preventing compiling.

dresdor commented:
Where there any additional changes that you wanted from this?

stajs commented:
Gah! My apologies. I thought I had reviewed this already, but I had failed to submit it meaning the above comment has been pending/hidden for a week.

![image](https://user-images.githubusercontent.com/2253814/35552644-9ce7f132-05f9-11e8-9be5-b30b36392c1a.png)

Just a minor fix and should be good to 🚢 !

dresdor committed:
Update Args.cs

dresdor commented:
Fixed.  Pretty stupid copy paste error since git was not cooperating with my VS.

stajs commented:
Thanks for your patience with seeing this through!
